processing/src/main/java/org/apache/druid/segment/join/Joinables.java

```diff
@@ -118,6 +107,68 @@ public static boolean isPrefixedBy(final String columnName, final String prefix)
     );
   }
 
+  /**
+   * Compute a cache key prefix for data sources that participate in the RHS of a join. This key prefix
+   * can be used in segment level cache or result level cache. The function can return following wrapped in an
+   * Optional
+   *  - Non-empty byte array - If there is join datasource involved and caching is possible. The result includes
+   *  join condition expression, join type and cache key returned by joinable factory for each {@link PreJoinableClause}
+   *  - NULL - There is a join but caching is not possible. It may happen if one of the participating datasource
+   *  in the JOIN is not cacheable.
+   *
+   * @throws {@link IAE} if this operation is called on a non-join data source
+   * @param dataSourceAnalysis for the join datasource
+   * @return the optional cache key to be used as part of query cache key
+   */
+  public Optional<byte[]> computeJoinDataSourceCacheKey(
+      final DataSourceAnalysis dataSourceAnalysis
+  )
+  {
+    final List<PreJoinableClause> clauses = dataSourceAnalysis.getPreJoinableClauses();
+    if (clauses.isEmpty()) {
+      throw new IAE("No join clauses to build the cache key for data source [%s]", dataSourceAnalysis.getDataSource());
+    }
+
+    final CacheKeyBuilder keyBuilder;
+    keyBuilder = new CacheKeyBuilder(JOIN_OPERATION);
+    for (PreJoinableClause clause : clauses) {
+      Optional<byte[]> bytes = joinableFactory.computeJoinCacheKey(clause.getDataSource(), clause.getCondition());
+      if (!bytes.isPresent()) {
+        // Encountered a data source which didn't support cache yet
+        log.debug("skipping caching for join since [%s] does not support caching", clause.getDataSource());
+        return Optional.empty();
+      }
+      keyBuilder.appendByteArray(bytes.get());
+      keyBuilder.appendString(clause.getCondition().getOriginalExpression());
+      keyBuilder.appendString(clause.getJoinType().name());
```

Comment:
The clause's prefix is important too, because it controls the names of the columns.

Reply:
functionally it makes no difference, right? do the cached results store column names with the prefix?

Reply:
It makes a difference because the prefixed fields might be inputs to aggregators, the group by clause, etc. If the prefix changes then those things would behave differently.

Reply:
added the prefix.